### PR TITLE
forest: parametrize Node over G: GrammarReflector instead of P aka G::NodeKind.

### PR DESCRIPTION
**Pull request by eddyb**
*Based on #13.*

This will get rid of repeated bounds on `P`, at the cost of having to manually write out impls on `Node` and having `Node<G>` be invariant over `G` (but the `ParseForest<G, I>` already is too anyway).

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -184,11 +176,11 @@ where
     }
 
     // FIXME(eddyb) safeguard this against misuse.
-    pub fn forest_add_split(&mut self, kind: P, left: Node<'i, P>) {
+    pub fn forest_add_split(&mut self, kind: G::NodeKind, left: Node<'i, G>) {
```
Comment (Centril) [inline src/parser.rs:179]: Btw, this method largely has the same structure as the one above aside from the first statement and the argument to `.insert` -- refactor?

File: src/forest.rs [hunk 2]
```diff
@@ -170,14 +199,14 @@ where
...
+        node: Node<'i, G>,
+    ) -> impl Iterator<Item = (Node<'i, G>, Node<'i, G>)> + Clone + 'a
     where
-        P: 'a,
+        G::NodeKind: 'a,
     {
         match self.grammar.node_shape(node.kind) {
             NodeShape::Split(left_kind, right_kind) => self
-                .possible_splits
+                .possibilities
```
Comment (Centril) [inline src/forest.rs]: The structure of this, aside from the lambda in `.map(...)` looks as the one in `all_choices` -- refactor?
There's also some duplication with `one_split`.

File: src/forest.rs [hunk 3]
```diff
@@ -124,30 +153,30 @@ where
...
+            NodeShape::Choice(_) => self
+                .possibilities
                 .get(&node)
                 .into_iter()
                 .flatten()
                 .cloned()
-                .map(move |kind| Node {
-                    kind,
+                .map(move |choice| Node {
+                    kind: self.grammar.node_shape_choice_get(node.kind, choice),
```
Comment (Centril) [inline src/forest.rs]: This bit here `Node { ... }` is repeated in `one_choice` -- refactor?

File: src/forest.rs [hunk 4]
```diff
@@ -47,19 +48,54 @@ impl<P> NodeShape<P> {
...
+where
+    G::NodeKind: PartialOrd,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        (self.kind, self.range).partial_cmp(&(other.kind, other.range))
+    }
+}
+impl<G: GrammarReflector> Ord for Node<'_, G>
+where
+    G::NodeKind: Ord,
```
Comment (Centril) [inline src/forest.rs:86]: Looks like ATBs need to become stable sooner =P

File: src/parser.rs (hunk 1 above)
Comment (eddyb) [reply src/parser.rs:179]: These are intentionally kept separate (tempted to rename `split`/`choice` to `product`/`sum`, if that helps).

File: src/forest.rs (hunk 2 above)
Comment (eddyb) [reply src/forest.rs]: (I agree, but this pertains to #13)

File: src/forest.rs (hunk 3 above)
Comment (eddyb) [reply src/forest.rs]: (I agree, but this pertains to #13)

**Review by CAD97 — APPROVED** [review]
Incremental diff is good :shipit:

**Pull request merged**